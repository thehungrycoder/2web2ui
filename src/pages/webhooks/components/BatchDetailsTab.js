import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { getBatches } from 'src/actions/webhooks';
import { selectWebhookBatches } from 'src/selectors/webhooks';
import { Panel, Tag, Tooltip, Table } from '@sparkpost/matchbox';
import { ArrowBack, InfoOutline } from '@sparkpost/matchbox-icons';
import { LabelledValue, PageLink } from 'src/components';
import _ from 'lodash';
import styles from './BatchDetailsTab.module.scss';

class BatchDetailsTab extends Component {
  componentDidMount() {
    this.getAttempts();
  }

  getAttempts = () => {
    const { webhook, getBatches, match } = this.props;
    const { id, subaccount } = webhook;
    return getBatches({ id, subaccount, params: { batch_ids: match.params.batchId }});
  }

  render() {
    const { batches, firstBatch = {}, webhook, loading } = this.props;
    const refreshAction = {
      content: loading ? 'Loading...' : 'Refresh Batches',
      color: 'orange',
      onClick: this.getAttempts,
      disabled: !!loading
    };

    return (
      <div>
        <Panel>
          <Panel.Section>
            <div className={styles.BackLink}><small><PageLink to={`/webhooks/details/${webhook.id}/batches`}><ArrowBack/> Return to All Batches</PageLink></small></div>
            <LabelledValue label='Batch ID'><h6>{firstBatch.batch_id}</h6></LabelledValue>
            <LabelledValue label='Attempts'><h6>{firstBatch.attempts}</h6></LabelledValue>
            <LabelledValue label='Batch Size'>
              <h6><Tooltip dark content='Number of events in this batch'>{firstBatch.batch_size} <InfoOutline className={styles.TooltipTrigger}/></Tooltip></h6>
            </LabelledValue>
          </Panel.Section>
        </Panel>
        <Panel title='Batch Attempt History' actions={[refreshAction]}>
          <Table>
            <thead>
              <Table.Row>
                <Table.HeaderCell width='200px'>Status</Table.HeaderCell>
                <Table.HeaderCell width='250px'>Timestamp</Table.HeaderCell>
                <Table.HeaderCell width='200px'>Latency</Table.HeaderCell>
                <Table.HeaderCell>Attempt</Table.HeaderCell>
              </Table.Row>
            </thead>
            <tbody>
              {batches.map((batch, i) => (
                <Table.Row key={i}>
                  <Table.Cell>
                    <Tag color={batch.status === 'Success' ? 'blue' : 'yellow'}>{batch.response_code}</Tag>
                  </Table.Cell>
                  <Table.Cell>{batch.formatted_time}</Table.Cell>
                  <Table.Cell>{batch.latency && `${batch.latency} ms`}</Table.Cell>
                  <Table.Cell>{batch.attempts}</Table.Cell>
                </Table.Row>
              ))}
            </tbody>
          </Table>
        </Panel>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  const batches = selectWebhookBatches(state);
  return {
    batches,
    firstBatch: _.first(batches),
    loading: state.webhooks.batchesLoading
  };
};

export default withRouter(connect(mapStateToProps, { getBatches })(BatchDetailsTab));
