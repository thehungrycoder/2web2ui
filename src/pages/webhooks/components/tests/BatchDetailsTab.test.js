import { shallow } from 'enzyme';
import React from 'react';

import { BatchDetailsTab } from '../BatchDetailsTab';

describe('Webhook Component: Batch Status Tab', () => {
  let wrapper;
  let props;

  beforeEach(() => {
    props = {
      getBatches: jest.fn(() => Promise.resolve()),
      webhook: {
        id: 'webhook-id',
        subaccount: 101
      },
      batchId: 'batch_1',
      batches: [
        {
          formatted_time: 'so-formatted',
          batch_id: 'batch_1',
          status: 'Success',
          attempts: 2,
          response_code: 200
        },
        {
          formatted_time: 'so-formatted-2',
          batch_id: 'batch_1',
          status: 'f',
          attempts: 1,
          response_code: 500
        }
      ],
      loading: false,
      firstBatch: {
        formatted_time: 'so-formatted',
        batch_id: '243423423423',
        status: 'Success',
        attempts: 1,
        response_code: 200
      }
    };

    wrapper = shallow(<BatchDetailsTab {...props} />);
  });

  it('should render batch details', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should get batches on mount', () => {
    expect(props.getBatches.mock.calls).toMatchSnapshot();
  });

  it('should handle loading state', () => {
    wrapper.setProps({ loading: true });
    expect(wrapper).toMatchSnapshot();
  });
});
