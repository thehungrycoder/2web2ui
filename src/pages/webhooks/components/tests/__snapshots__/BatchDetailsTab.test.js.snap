// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Webhook Component: Batch Status Tab should get batches on mount 1`] = `
Array [
  Array [
    Object {
      "id": "webhook-id",
      "params": Object {
        "batch_ids": "batch_1",
      },
      "subaccount": 101,
    },
  ],
]
`;

exports[`Webhook Component: Batch Status Tab should handle loading state 1`] = `
<div>
  <Panel>
    <Panel.Section>
      <div
        className="BackLink"
      >
        <small>
          <PageLink
            to="/webhooks/details/webhook-id/batches"
          >
            <ArrowBack />
             Return to All Batches
          </PageLink>
        </small>
      </div>
      <LabelledValue
        label="Batch ID"
      >
        <h6>
          243423423423
        </h6>
      </LabelledValue>
      <LabelledValue
        label="Attempts"
      >
        <h6>
          1
        </h6>
      </LabelledValue>
      <LabelledValue
        label="Batch Size"
      >
        <h6>
          <Tooltip
            bottom={true}
            content="Number of events in this batch"
            dark={true}
            forcePosition={false}
            horizontalOffset="0px"
            preferredDirection={
              Object {
                "bottom": true,
                "left": false,
                "right": true,
                "top": false,
              }
            }
            right={true}
          >
             
            <InfoOutline
              className="TooltipTrigger"
            />
          </Tooltip>
        </h6>
      </LabelledValue>
    </Panel.Section>
  </Panel>
  <Panel
    actions={
      Array [
        Object {
          "color": "orange",
          "content": "Loading...",
          "disabled": true,
          "onClick": [Function],
        },
      ]
    }
    title="Batch Attempt History"
  >
    <Table>
      <thead>
        <Table.Row>
          <Table.HeaderCell
            width="200px"
          >
            Status
          </Table.HeaderCell>
          <Table.HeaderCell
            width="250px"
          >
            Timestamp
          </Table.HeaderCell>
          <Table.HeaderCell
            width="200px"
          >
            Latency
          </Table.HeaderCell>
          <Table.HeaderCell>
            Attempt
          </Table.HeaderCell>
        </Table.Row>
      </thead>
      <tbody>
        <Table.Row
          key="0"
        >
          <Table.Cell>
            <Tag
              color="blue"
            >
              200
            </Tag>
          </Table.Cell>
          <Table.Cell>
            so-formatted
          </Table.Cell>
          <Table.Cell />
          <Table.Cell>
            2
          </Table.Cell>
        </Table.Row>
        <Table.Row
          key="1"
        >
          <Table.Cell>
            <Tag
              color="yellow"
            >
              500
            </Tag>
          </Table.Cell>
          <Table.Cell>
            so-formatted-2
          </Table.Cell>
          <Table.Cell />
          <Table.Cell>
            1
          </Table.Cell>
        </Table.Row>
      </tbody>
    </Table>
  </Panel>
</div>
`;

exports[`Webhook Component: Batch Status Tab should render batch details 1`] = `
<div>
  <Panel>
    <Panel.Section>
      <div
        className="BackLink"
      >
        <small>
          <PageLink
            to="/webhooks/details/webhook-id/batches"
          >
            <ArrowBack />
             Return to All Batches
          </PageLink>
        </small>
      </div>
      <LabelledValue
        label="Batch ID"
      >
        <h6>
          243423423423
        </h6>
      </LabelledValue>
      <LabelledValue
        label="Attempts"
      >
        <h6>
          1
        </h6>
      </LabelledValue>
      <LabelledValue
        label="Batch Size"
      >
        <h6>
          <Tooltip
            bottom={true}
            content="Number of events in this batch"
            dark={true}
            forcePosition={false}
            horizontalOffset="0px"
            preferredDirection={
              Object {
                "bottom": true,
                "left": false,
                "right": true,
                "top": false,
              }
            }
            right={true}
          >
             
            <InfoOutline
              className="TooltipTrigger"
            />
          </Tooltip>
        </h6>
      </LabelledValue>
    </Panel.Section>
  </Panel>
  <Panel
    actions={
      Array [
        Object {
          "color": "orange",
          "content": "Refresh Batches",
          "disabled": false,
          "onClick": [Function],
        },
      ]
    }
    title="Batch Attempt History"
  >
    <Table>
      <thead>
        <Table.Row>
          <Table.HeaderCell
            width="200px"
          >
            Status
          </Table.HeaderCell>
          <Table.HeaderCell
            width="250px"
          >
            Timestamp
          </Table.HeaderCell>
          <Table.HeaderCell
            width="200px"
          >
            Latency
          </Table.HeaderCell>
          <Table.HeaderCell>
            Attempt
          </Table.HeaderCell>
        </Table.Row>
      </thead>
      <tbody>
        <Table.Row
          key="0"
        >
          <Table.Cell>
            <Tag
              color="blue"
            >
              200
            </Tag>
          </Table.Cell>
          <Table.Cell>
            so-formatted
          </Table.Cell>
          <Table.Cell />
          <Table.Cell>
            2
          </Table.Cell>
        </Table.Row>
        <Table.Row
          key="1"
        >
          <Table.Cell>
            <Tag
              color="yellow"
            >
              500
            </Tag>
          </Table.Cell>
          <Table.Cell>
            so-formatted-2
          </Table.Cell>
          <Table.Cell />
          <Table.Cell>
            1
          </Table.Cell>
        </Table.Row>
      </tbody>
    </Table>
  </Panel>
</div>
`;
